lib/FrontendTool/FrontendTool.cpp

```diff
@@ -262,6 +264,313 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+        addToABIExportMap(module, reexposedImport);
+      }
+    }
+  }
+
+  SmallPtrSet<ModuleDecl *, 32> reexportedImports;
+  ::getImmediateImports(module, reexportedImports,
+                        {ModuleDecl::ImportFilterKind::Public});
+  for (auto reexportedImport: reexportedImports) {
+    computeABIDependenciesForModule(reexportedImport);
+    reexposeImportedABI(module, reexportedImport);
+  }
+}
+
+void ABIDependencyEvaluator::computeABIDependenciesForClangModule(
+    ModuleDecl *module) {
+  SmallPtrSet<ModuleDecl *, 32> imports;
+  ::getImmediateImports(module, imports);
+  for (auto import: imports) {
+    // There are three cases here which can potentially create cycles:
+    //
+    // 1. Clang modules importing the stdlib.
+    //    See [Note: Pure-Clang-modules-privately-import-stdlib].
+    // 2. Overlay S @_exported-imports underlying module S' and another Clang
+    //    module C'. C' (transitively) #imports S' but it gets treated as if
+    //    C' imports S. This creates a cycle: S -> C' -> ... -> S.
+    //    In practice, this case is hit for
+    //      Darwin (Swift) -> SwiftOverlayShims (Clang) -> Darwin (Swift).
+    // 3. [Note: Intra-module-leafwards-traversal]
+    //    Cycles within the same top-level module.
+    //    These don't matter for us, since we only care about the dependency
+    //    graph at the granularity of top-level modules. So we ignore these
+    //    by only considering parent -> submodule dependencies.
+    //    See also [Note: Trace-Clang-submodule-complexity].
+    if (import->isStdlibModule()) {
+      continue;
+    }
+    if (!import->isNonSwiftModule()
+        && import->findUnderlyingClangModule() != nullptr
+        && llvm::find(searchStack, import) != searchStack.end()) {
+      continue;
+    }
+    if (import->isNonSwiftModule()
+        && module->getTopLevelModule() == import->getTopLevelModule()
+        && !import->findUnderlyingClangModule()
+                  ->isSubModuleOf(module->findUnderlyingClangModule())) {
+      continue;
+    }
+    computeABIDependenciesForModule(import);
+    reexposeImportedABI(module, import);
+  }
+}
+
+void ABIDependencyEvaluator::getABIDependenciesForSwiftModule(
+    ModuleDecl *module, SmallPtrSetImpl<ModuleDecl *> &abiDependencies) {
+  computeABIDependenciesForModule(module);
+  SmallPtrSet<ModuleDecl *, 32> allImports;
+  ::getImmediateImports(module, allImports);
+  for (auto directDependency: allImports) {
+    abiDependencies.insert(directDependency);
+    for (auto exposedDependency: abiExportMap[directDependency]) {
+      abiDependencies.insert(exposedDependency);
+    }
+  }
+}
+
+void ABIDependencyEvaluator::printModule(
+    const ModuleDecl *module, llvm::raw_ostream &os) {
+  module->getReverseFullModuleName().printForward(os);
+  os << (module->isNonSwiftModule() ? " (Clang)" : " (Swift)");
+  os << " @ " << llvm::format("0x%llx", reinterpret_cast<uintptr_t>(module));
```

Comment:
This might seem a bit odd but is handy for confirming uniqueness by looking at the pointer values.